README.md
<div align="center">
  
[![npm](https://badgen.net/npm/v/rn-segmented-control)](https://www.npmjs.com/package/rn-segmented-control) [![npm](https://badgen.net/npm/dt/rn-segmented-control)](https://www.npmjs.com/package/rn-segmented-control) [![npm](https://badgen.net/npm/license/rn-segmented-control)](https://www.npmjs.com/package/rn-segmented-control) [![](https://badgen.net/npm/types/tslib)](https://badgen.net/npm/types/tslib)
  
<h1>React Native Segmented Control</h1>

<img width="auto" height="600" src="./examples/RNSegmentedControl/assets/segmentcontrolios.gif">

<img width="auto" height="600" src="./examples/RNSegmentedControl/assets/segmentcontroliosdark.gif">

</div>

---

## :anchor: Installation

```sh
yarn add rn-segmented-control
# or
npm i rn-segmented-control
```

## :arrows_clockwise: Dependencies

```sh
npm i prop-types
```

## :mag: Usage

```js
import React from "react";
import { Dimensions, StyleSheet, Text, View, Button } from "react-native";
import SegmentedControl from "rn-segmented-control";

const AppRoot = () => {
  const [tabIndex, setTabIndex] = React.useState(1);
  const [theme, setTheme] = React.useState("LIGHT");
  const toggleTheme = () =>
    theme === "LIGHT" ? setTheme("DARK") : setTheme("LIGHT");
  const handleTabsChange = (index) => {
    setTabIndex(index);
  };
  return (
    <View
      style={[
        styles.container,
        { backgroundColor: theme === "LIGHT" ? "white" : "black" },
      ]}
    >
      <Text
        style={[
          styles.textStyle,
          { color: theme === "LIGHT" ? "black" : "white" },
        ]}
      >
        Hello,World !
      </Text>
      <Button title="Toggle Theme" onPress={toggleTheme}></Button>
      <Text
        style={[
          styles.textStyle,
          { color: theme === "LIGHT" ? "black" : "white" },
        ]}
      >
        Segmented Control with 2 labels
      </Text>
      {/* Default Segmented Control */}
      <SegmentedControl
        tabs={["Label", "Label"]}
        onChange={() => {}}
        paddingVertical={6}
        containerStyle={{
          marginVertical: 20,
        }}
        currentIndex={tabIndex}
        onChange={handleTabsChange}
        theme={theme}
      />
      <Text
        style={[
          styles.textStyle,
          { color: theme === "LIGHT" ? "black" : "white" },
        ]}
      >
        {" "}
        Segmented Control with 3 labels
      </Text>
      {/* Segmented Control with a custom width */}
      <SegmentedControl
        tabs={["Label", "Label", "Label"]}
        onChange={() => {}}
        paddingVertical={10}
        containerStyle={{
          marginVertical: 20,
        }}
        width={Dimensions.get("screen").width - 200}
        theme={theme}
      />
      <Text
        style={[
          styles.textStyle,
          { color: theme === "LIGHT" ? "black" : "white" },
        ]}
      >
        {" "}
        Segmented Control with 4 labels
      </Text>
      <SegmentedControl
        tabs={["Label", "Label", "Label", "Label"]}
        onChange={() => {}}
        paddingVertical={14}
        width={Dimensions.get("screen").width - 90}
        containerStyle={{
          marginVertical: 20,
        }}
        textStyle={{
          fontWeight: "300",
          fontSize: 24,
        }}
        theme={theme}
      />
      <Text
        style={[
          styles.textStyle,
          { color: theme === "LIGHT" ? "black" : "white" },
        ]}
      >
        Customised Segmented Control
      </Text>
      {/* Segmented Control with a custom width, container style and font weight */}
      <SegmentedControl
        tabs={["Shop", "Explore", "Search"]}
        segmentedControlBackgroundColor="#86c4fD"
        activeSegmentBackgroundColor="#0482f7"
        activeTextColor="white"
        textColor="black"
        paddingVertical={10}
        width={Dimensions.get("screen").width - 100}
        containerStyle={{
          marginVertical: 20,
        }}
        textStyle={{
          fontWeight: "300",
        }}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    paddingHorizontal: 16,
  },
  textStyle: {
    fontSize: 24,
    textAlign: "center",
    paddingVertical: 10,
  },
});

export default AppRoot;
```

## :camera: Screenshot

| iOS                                                                                            | Android                                                                                            | Dark Mode                                                                                          |
| ---------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------- |
| <img width="auto" height="500" src='./examples/RNSegmentedControl/assets/screenshotios.png' /> | <img width="auto" height="500" src='./examples/RNSegmentedControl/assets/screenshotandroid.png' /> | <img width="auto" height="500" src='./examples/RNSegmentedControl/assets/screenshotiosdark.png' /> |

## :wrench: Props

| Name                            | Description                                    | Required | Type                 | Default               |
| ------------------------------- | ---------------------------------------------- | -------- | -------------------- | --------------------- |
| tabs                            | An array of labels for segments                | YES      | Array                | []                    |
| onChange                        | A callback Function with pressed segment index | YES      | Function             | () => {}              |
| currentIndex                    | Index for the currently active segment         | YES      | Number               | 0                     |
| activeSegmentBackgroundColor    | Background color of Active Label               | NO       | Color                | Set by theme          |
| segmentedControlBackgroundColor | Background color of the segment                | NO       | Color                | Set by theme          |
| textColor                       | Color of Label Text                            | NO       | Color                | Set by theme          |
| activeTextColor                 | Color of Active Label Text                     | NO       | Color                | Set by theme          |
| activeTextWeight                | Weight of Active Label Text                    | NO       | String               | 600                   |
| paddingVertical                 | A numeric value to manually adjust the height. | NO       | Number               | 12                    |
| width                           | Sets the width of the segmented control        | NO       | Number               | Screen Specific Width |
| containerStyle                  | Style object for the Segmented Control         | NO       | ViewPropTypes.style  | {}                    |
| tileStyle                       | Style object for the absolute positioned tile  | NO       | ViewPropTypes.style  | {}                    |
| textStyle                       | Style object for the Labels                    | NO       | Text.style           | {}                    |
| isRTL                           | Controls the toggle animation direction        | NO       | Bool                 | false                 |
| theme                           | App Theme                                      | NO       | oneOf['LIGHT','DARK] | 'LIGHT'               |
| shadowStyle                     | Style for segment background                   | NO       | Object`[1]`          | null                  |

```js
   [1]
   Shadow Style
   PropTypes.shape({
    shadowColor: PropTypes.string,
    shadowOffset: PropTypes.shape({
      width: PropTypes.number,
      height: PropTypes.number,
    }),
    shadowOpacity: PropTypes.number,
    shadowRadius: PropTypes.number,
    elevation: PropTypes.number,
  })
```

> Color accepts all [React Native Colors](https://reactnative.dev/docs/colors#color-representations).

> :warning: `containerStyle` accepts all View styles and does override some default styles provided by the package. Make sure you use it properly :)

> `textStyle` accepts all Text styles and overrides the defaults except the label colors set by props

> :warning: `tileStyle` accepts all View styles and does override some default styles provided by the package. Make sure you use it properly :)

## :tada: Example

Checkout the example [here](https://github.com/Karthik-B-06/rn-segmented-control/tree/master/examples/RNSegmentedControl).

## :notebook: Blog

Have a look at my blog [here](https://medium.com/timeless/react-native-segmented-control-92508dcba97c).

## :snowman: Built with ❤️

- [react-native](https://www.npmjs.com/package/react-native)

## :white_check_mark: Todos

- [ ] Custom Animation Spring Config
- [ ] Add React Native Reanimated library for animations

## :v: Contributing

Pull requests are always welcome! Feel free to open a new GitHub issue for any changes that can be made.

## :man: Author

[Karthik B](https://twitter.com/_iam_karthik)

## :clipboard: License

MIT

---

<div align='center'>
  <h4>   Like this project ?? </h4>
<a href="https://www.buymeacoffee.com/karthik06" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-blue.png" alt="Buy Me A Coffee" height="50" ></a>
</div>

---


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.2.9](https://github.com/Karthik-B-06/rn-segmented-control/compare/v0.2.8...v0.2.9) (2021-12-12)


### Features

* :memo: Updated README ([67106a2](https://github.com/Karthik-B-06/rn-segmented-control/commit/67106a28b27f3ec752b184c430f2167d1e059173))
* :memo: Updated README with new prop desc ([aa8ad6d](https://github.com/Karthik-B-06/rn-segmented-control/commit/aa8ad6db121e69a12113f5fe3a96734d63372dcb))

### [0.2.8](https://github.com/Karthik-B-06/rn-segmented-control/compare/v0.2.7...v0.2.8) (2021-04-24)


### Features

* :label: Added new prop `tileStyle` ([d668d93](https://github.com/Karthik-B-06/rn-segmented-control/commit/d668d938b8d6dd85878cd7b8188b34caeab052e0))
* :rocket: Created ios build with pod install ([ab0b630](https://github.com/Karthik-B-06/rn-segmented-control/commit/ab0b630ada41c4b785a676907a7d96f9474d8978))
* :zap: Added example with custom tile style and fixing style overriding issue of [#18](https://github.com/Karthik-B-06/rn-segmented-control/issues/18) ([d5838d7](https://github.com/Karthik-B-06/rn-segmented-control/commit/d5838d7fb335247b4f60c955af56ba37cc07dc4a))
* :zap: Fixed style overriding through containerStyle prop and added new prop tileStyle ([1a805f1](https://github.com/Karthik-B-06/rn-segmented-control/commit/1a805f1c9268f3365e9be120870dc1b93ca3a678))

### 0.2.7 (2020-10-24)

- :fire: Added type definitions

### Features

- :fire: Removed yarn.lock ([2634cb1](https://github.com/Karthik-B-06/rn-segmented-control/commit/2634cb1a5c2dab10cce4539d08f454c761942f99))
- :rocket: version bumped to 0.2.6 ([af938ba](https://github.com/Karthik-B-06/rn-segmented-control/commit/af938ba640b148c4bef0a0f7ab5f17f6385b3d66))
- Added standard-version ([fa799e4](https://github.com/Karthik-B-06/rn-segmented-control/commit/fa799e46124f0f41a39425b6117e408cd7e38056))
- Added type definitions ([4989b25](https://github.com/Karthik-B-06/rn-segmented-control/commit/4989b25c4fbc189de01f2a702c20bd4e91038fac))

### Bug Fixes

- :ambulance: Upgrade node-fetch to version 2.6.1 ([6fabecf](https://github.com/Karthik-B-06/rn-segmented-control/commit/6fabecfda9b1b37d8c940d5cc1389c1393d72aca))
- :bug: Added default weight for activeTextWeight prop ([34605ba](https://github.com/Karthik-B-06/rn-segmented-control/commit/34605ba7a965a0ef54016caf786b0a6e27cad7dd))
- :rotating_light: Fixed prop type shadowStyle is invalid warning ([2697e92](https://github.com/Karthik-B-06/rn-segmented-control/commit/2697e924944b295f7168ea99564d7072c20d7d3a))
- :rotating_light: Fixed prop type shadowStyle is invalid warning ([830576b](https://github.com/Karthik-B-06/rn-segmented-control/commit/830576b8ef81199626231e1d021a13384ac9b306))


src/segmentedControl/index.js
import PropTypes from 'prop-types';
import React, { useEffect } from 'react';
import { Animated, Dimensions, StyleSheet, Text, TouchableOpacity, ViewPropTypes } from 'react-native';

const getSegmentedBackgroundColor = (theme, colorValueFromProps) => {
  return colorValueFromProps || (theme === 'LIGHT' ? '#E5E5EA' : '#4a5568');
};

const getSegmentedTextColor = (theme, colorValueFromProps) => {
  return colorValueFromProps || (theme === 'LIGHT' ? 'black' : 'white');
};

const getActiveSegmentedBackgroundColor = (theme, colorValueFromProps) => {
  return colorValueFromProps || (theme === 'LIGHT' ? 'white' : 'black');
};

const getActiveSegmentedTextColor = (theme, colorValueFromProps) => {
  return colorValueFromProps || (theme === 'LIGHT' ? 'black' : 'white');
};

const defaultShadowStyle = {
  shadowColor: '#000',
  shadowOffset: {
    width: 0,
    height: 2
  },
  shadowOpacity: 0.23,
  shadowRadius: 2.62,

  elevation: 4
};

const SegmentedControl = (props) => {
  const { width, shadowStyle } = props;
  const marginHorizontal = (props.tileStyle.marginHorizontal || 2) * 2;
  const translateValue = (width - marginHorizontal) / props?.tabs?.length;
  const [tabTranslate] = React.useState(new Animated.Value(0));
  const shadow = shadowStyle || defaultShadowStyle;
  // useCallBack with an empty array as input, which will call inner lambda only once and memoize the reference for future calls
  const memoizedTabPressCallback = React.useCallback(
    (index) => {
      props?.onChange(index);
    },
    [props?.onChange]
  );

  useEffect(() => {
    // If phone is set to RTL, make sure the animation does the correct transition.
    const transitionMultiplier = props?.isRTL ? -1 : 1;

    // Animating the active index based current index
    Animated.spring(tabTranslate, {
      toValue: props?.currentIndex * (transitionMultiplier * translateValue),
      stiffness: 150,
      damping: 20,
      mass: 1,
      useNativeDriver: true
    }).start();
  }, [props?.currentIndex]);

  return (
    <Animated.View
      style={[
        styles.segmentedControlWrapper,
        props?.containerStyle,
        {
          width: width
        },
        {
          backgroundColor: getSegmentedBackgroundColor(props?.theme, props?.segmentedControlBackgroundColor)
        },
        {
          paddingVertical: props?.paddingVertical
        }
      ]}
    >
      <Animated.View
        style={[
          styles.defaultTileStyle,
          props?.tileStyle,
          {
            ...StyleSheet.absoluteFill,
            position: 'absolute',
            width: (width - marginHorizontal) / props?.tabs?.length,
            top: 0,
            backgroundColor: getActiveSegmentedBackgroundColor(props?.theme, props?.activeSegmentBackgroundColor),
            ...shadow
          },
          {
            transform: [
              {
                translateX: tabTranslate
              }
            ]
          }
        ]}
      ></Animated.View>
      {props?.tabs.map((tab, index) => {
        const isCurrentIndex = props?.currentIndex === index;
        return (
          <TouchableOpacity
            key={index}
            style={[styles.textWrapper]}
            activeOpacity={0.7}
            onPress={() => memoizedTabPressCallback(index)}
          >
            <Text
              numberOfLines={1}
              style={[
                styles.textStyles,
                props?.textStyle,
                isCurrentIndex
                  ? {
                    color: getActiveSegmentedTextColor(props?.theme, props?.activeTextColor),
                    fontWeight: props?.activeTextWeight
                  }
                  : {
                    color: getSegmentedTextColor(props?.theme, props?.textColor)
                  }
              ]}
            >
              {tab}
            </Text>
          </TouchableOpacity>
        );
      })}
    </Animated.View>
  );
};

const styles = StyleSheet.create({
  segmentedControlWrapper: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    borderRadius: 8
  },
  textWrapper: {
    flex: 1,
    elevation: 9,
    paddingHorizontal: 5
  },
  textStyles: {
    fontSize: 18,
    textAlign: 'center',
    fontWeight: '600'
  },
  defaultTileStyle: {
    marginVertical: 2,
    marginHorizontal: 2,
    borderRadius: 8
  }
});

SegmentedControl.propTypes = {
  tabs: PropTypes.arrayOf(PropTypes.string).isRequired,
  onChange: PropTypes.func.isRequired,
  currentIndex: PropTypes.number.isRequired,
  segmentedControlBackgroundColor: PropTypes.string,
  activeSegmentBackgroundColor: PropTypes.string,
  textColor: PropTypes.string,
  activeTextColor: PropTypes.string,
  activeTextWeight: PropTypes.string,
  paddingVertical: PropTypes.number,
  width: PropTypes.number,
  containerStyle: ViewPropTypes.style,
  textStyle: PropTypes.object,
  isRTL: PropTypes.bool,
  theme: PropTypes.oneOf(['LIGHT', 'DARK']),
  shadowStyle: PropTypes.shape({
    shadowColor: PropTypes.string,
    shadowOffset: PropTypes.shape({
      width: PropTypes.number,
      height: PropTypes.number
    }),
    shadowOpacity: PropTypes.number,
    shadowRadius: PropTypes.number,
    elevation: PropTypes.number
  }),
  tileStyle: ViewPropTypes.style
};

SegmentedControl.defaultProps = {
  tabs: [],
  onChange: () => {},
  currentIndex: 0,
  segmentedControlBackgroundColor: null,
  activeSegmentBackgroundColor: null,
  textColor: null,
  activeTextColor: null,
  activeTextWeight: '600',
  paddingVertical: 12,
  width: Dimensions.get('screen').width - 32,
  containerStyle: {},
  textStyle: {},
  isRTL: false,
  theme: 'LIGHT',
  shadowStyle: null,
  tileStyle: {}
};

export default SegmentedControl;
